### ISSUE TITLE
iOS Crash because of using OpenGL when App is in background.

### ISSUE THREAD

**Issue by wqyfavor** [flutter/flutter#57464]
Steps to reproduce.

Using demo with a list of images. You can get the Dart source as attached file.

Run on device with older iOS versions, my device is iOS9.

Debug in Xcode and after App launches quickly press home button to switch the app to background.

It crashed.

Source code
```dart
import 'dart:math';

import 'package:flutter/material.dart';

void main() => runApp(new MyApp());

var COLORS = [
  Color(0xFFEF7A85),
  Color(0xFFFF90B3),
  Color(0xFFFFC2E2),
  Color(0xFFB892FF),
  Color(0xFFB892FF)
];

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      debugShowCheckedModeBanner: false,
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'Awesome List'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  var data = [
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    },
    {
      "title": "Hey Flutterers, See what I did with Flutter",
      "content": "This is just a text description of the item",
      "color": COLORS[new Random().nextInt(5)],
      "image": "https://picsum.photos/1000?random"
    }
  ];

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      backgroundColor: Colors.white,
      appBar: new AppBar(
        backgroundColor: Colors.transparent,
        elevation: 0.0,
        title: new Text(
          widget.title,
          style: TextStyle(color: Colors.white),
        ),
      ),
      body: ListView.builder(
        shrinkWrap: true,
        padding: const EdgeInsets.all(0.0),
        scrollDirection: Axis.vertical,
        primary: true,
        itemCount: data.length,
        itemBuilder: (BuildContext content, int index) {
          return AwesomeListItem(
              title: data[index]["title"],
              content: data[index]["content"],
              color: data[index]["color"],
              image: data[index]["image"],
              index: index + 1);
        },
      ),
    );
  }
}

class MyClipper extends CustomClipper<Path> {
  @override
  Path getClip(Size size) {
    Path p = new Path();
    p.lineTo(size.width, 0.0);
    p.lineTo(size.width, size.height / 4.75);
    p.lineTo(0.0, size.height / 3.75);
    p.close();
    return p;
  }

  @override
  bool shouldReclip(CustomClipper oldClipper) {
    return true;
  }
}

class AwesomeListItem extends StatefulWidget {
  String title;
  String content;
  Color color;
  String image;
  int index;

  AwesomeListItem({this.title, this.content, this.color, this.image, this.index});

  @override
  _AwesomeListItemState createState() => new _AwesomeListItemState();
}

class _AwesomeListItemState extends State<AwesomeListItem> {
  @override
  Widget build(BuildContext context) {
    return new Container(
      height: 375,
      width: 375,
      color: Colors.red,
      child: Image.network('https://i.picsum.photos/id/' + widget.index.toString() + '/1000/1000.jpg')
    );
  }
}
```

<img width="1286" alt="截屏2020-05-18 下午2 27 17" src="https://user-images.githubusercontent.com/113776/82180800-dc705880-9913-11ea-945c-7843a42ec283.png">

**Comment by VladyslavBondarenko** [comment]
Hi @wqyfavor 
could you provide your `flutter doctor -v` and error log from xcode?

**Comment by wqyfavor** [comment]
> Hi @wqyfavor
> could you provide your `flutter doctor -v` and error log from xcode?

stable/v1.12.13+hotfix.8

**Comment by wqyfavor** [comment]
Actually we found three cases which might produce this crash.

This is the process of debugging this problem.

1. We found our released App crashes
![截屏2020-05-19 上午10 48 45](https://user-images.githubusercontent.com/113776/82279116-565d1c00-99be-11ea-8966-49678dbe5bfd.png)

I guess that user switched to background immediately after starting the App and try to reproduce in a demo.

2. In the demo, we cannot reproduce user's crash. But found that `flutter::Picture::RasterizeToImage` crashes because of using OpenGL in background. Just as the image above shows.

3. I comment the code in `RasterizeToImage`and found another case of using OpenGL in background. It now crashes at image_decode.cc `UploadRasterImage`
<img width="1416" alt="截屏2020-05-19 上午10 56 32" src="https://user-images.githubusercontent.com/113776/82279586-7d681d80-99bf-11ea-8068-6f8cebdad10f.png">

I believe that's because the GPUDisabledSyncSwitch is false by default and it does not get chance to be changed to true when user quickly switches to background after App starts up. Actually I logged in `-[FlutterEngine applicationWillResignActive]`. It is not executed when App crashes so the switch is not changed to true.

4. I also comment the code of 'False' case in UploadRasterImage and always use MakeFromRaster to generate CPU image. Now the App crashes at `[[EAGLContext currentContext] presentRenderbuffer:GL_RENDERBUFFER]`

<img width="1357" alt="截屏2020-05-19 上午11 06 19" src="https://user-images.githubusercontent.com/113776/82280146-c9679200-99c0-11ea-9b3e-986786ffb5a9.png">

Finally. All the crashes can be reproduced in debug unoptimized mode debugging a demo app in Xcode on an iOS9 device.

For crash at `presentRenderbuffer` I use such code to fix. Only notify Dart 'resumed' when App is active.
<img width="1256" alt="截屏2020-05-19 上午11 08 42" src="https://user-images.githubusercontent.com/113776/82280292-2105fd80-99c1-11ea-8def-0081b97d83dc.png">

For crash at `UploadRasterImage`, my fix is that setting the `IsGpuDisabledSyncSwitch` to true by default and only allows uploading GPU texture when FlutterEngine definitely received `applicationBecameActive` notification.

But for crash at `RasterizeToImage`, I do not know the proper fix. After I fix `resumed` and `UploadRasterImage` the app still crashes at `RasterizeToImage`

**Comment by wqyfavor** [comment]
By the way, currently we use such code to protect RasterizeToImage. 
But the problem is that in Debug mode, `fml::WeakPtr<IOManager>->GetIsGpuDisabledSyncSwitch` is not allowed to be invoked in Dart.UI thread. But in release it would be OK.

<img width="855" alt="截屏2020-05-19 下午12 11 50" src="https://user-images.githubusercontent.com/113776/82283804-fc625380-99c9-11ea-8ebe-3b687775c4bf.png">

**Comment by kangwang1988** [comment]
@gaaclarke  
@wqyfavor is my colleague and trying to resolve libgpu related crash.
We're always using flutter in a add2app scenario.
In the past, the flutter page is not the first page, users may need to click something to enter a flutter page, in that case, the libgpu related crash has been reduced to a pretty low level in stable v1.12.13+hotfix.8, in other words, the syncswitch mechanism help.

However, now some teams are using flutter in the application where the flutter page would be launched in the very beginning. In this case, the fluttervc is the first item in a tabviewcontroller acting as the rootviewcontroller for the application. The libgpu crash is easily to reproduced when we press the home button when launching. The reason is that the UIViewController lifecycle might continue to run(gpu related operation would be triggered) before the notification based gpu operation control.

@wqyfavor 's approach seems working in our scenario, however, it would be better that you can review it and check if it needs to be considered in the upstream.

Also CC @xster @dnfield

**Comment by wqyfavor** [comment]
I will paste my fix here and we can discuss. Finally I can generate a pull request.

Fix of RasterizeToImage, which might cause FML_DCHECK_CREATION_THREAD_IS_CURRENT failure in debug mode. This needs discussion.
<img width="854" alt="截屏2020-05-19 下午6 13 39" src="https://user-images.githubusercontent.com/113776/82314486-7b707f80-99fc-11ea-93ac-dbe1ea2f2983.png">

Change sync switch to true by default.
<img width="616" alt="截屏2020-05-19 下午6 15 06" src="https://user-images.githubusercontent.com/113776/82314641-b2df2c00-99fc-11ea-99b7-d30a315aa2bc.png">

<img width="922" alt="截屏2020-05-19 下午6 15 55" src="https://user-images.githubusercontent.com/113776/82314748-d5714500-99fc-11ea-900b-f2ddba365860.png">

In viewDidAppear of FlutterViewController check App state.
<img width="735" alt="截屏2020-05-19 下午6 17 00" src="https://user-images.githubusercontent.com/113776/82314826-f5086d80-99fc-11ea-8e3b-b8b1d48ae1e5.png">

**Comment by gaaclarke** [comment]
The fix for backgrounding and OpenGL calls happened in [v1.16.3](https://github.com/flutter/flutter/commit/144eb45aa447bd7abde4f05cb24c6c13ced0971f).  Please try with that version or later.

**Comment by kangwang1988** [comment]
@gaaclarke 
As we can't upgrade to v1.17.1 in the first time, can you paste the relevant fix commit so that we can cherry-pick and reuse it in v1.12.13?

**Comment by wqyfavor** [comment]
@gaaclarke I could not find the commit fixing the problems I mentioned in this issue. Do you mean https://github.com/flutter/flutter/commit/97a23a80e1b4f744c6d9d9f6d83f5c0df1aeb11f ?

**Comment by gaaclarke** [comment]
@kangwang1988 @wqyfavor I linked the roll into Flutter, the actual engine commit is https://github.com/flutter/flutter/commit/97a23a80e1b4f744c6d9d9f6d83f5c0df1aeb11f.

**Comment by wqyfavor** [comment]
> @kangwang1988 @wqyfavor I linked the roll into Flutter, the actual engine commit is [97a23a8](https://github.com/flutter/flutter/commit/97a23a80e1b4f744c6d9d9f6d83f5c0df1aeb11f).

Hi, gaaclarke

Commit 97a23a8 does not fix the problem in this issue. The engine we use is v1.12.13 which already contains this commit. The three cases in this issue still yield background OpenGL calls.

**Comment by wqyfavor** [comment]
@gaaclarke

**Comment by kangwang1988** [comment]
> > @kangwang1988 @wqyfavor I linked the roll into Flutter, the actual engine commit is [97a23a8](https://github.com/flutter/flutter/commit/97a23a80e1b4f744c6d9d9f6d83f5c0df1aeb11f).
> 
> Hi, @gaaclarke
> 
> Commit [97a23a8](https://github.com/flutter/flutter/commit/97a23a80e1b4f744c6d9d9f6d83f5c0df1aeb11f) does not fix the problem in this issue. The engine we use is v1.12.13 which already contains this commit. The three cases in this issue still produce background OpenGL calls.

I'm going to reopen this issue as it seems this issue is not fixed.

**Comment by gaaclarke** [comment]
> The engine we use is v1.12.13 which already contains this commit. The three cases in this issue still produce background OpenGL calls.

@wqyfavor The change wasn't incorporated into Flutter until [v1.16.3](https://github.com/flutter/flutter/commit/144eb45aa447bd7abde4f05cb24c6c13ced0971f).  Please see the tag in that roll commit.  Just to be sure something weird didn't happen I just verified locally that that change does not exist in v1.12.13.

> The three cases in this issue still produce background OpenGL calls.

Are you testing with v1.12.13 or v1.16.3?

**Comment by gaaclarke** [comment]
@wqyfavor I can't reproduce the bug locally but you've provided enough information to feel pretty confident https://github.com/flutter/engine/pull/18563 fixes your issue.  Do you have a way that you can test that?  I'll try a few more ways to reproduce it.

**Comment by gaaclarke** [comment]
>I believe that's because the GPUDisabledSyncSwitch is false by default and it does not get chance to be changed to true when user quickly switches to background after App starts up. Actually I logged in -[FlutterEngine applicationWillResignActive]. It is not executed when App crashes so the switch is not changed to true.

This a risky change.  Crashing and not-crashing looks exactly the same in the case where a user backgrounds the app before `applicationBecameActive`.  It has zero effect on user experience.  I wouldn't recommend any change for that.

**Comment by wqyfavor** [comment]
> > The engine we use is v1.12.13 which already contains this commit. The three cases in this issue still produce background OpenGL calls.
> 
> @wqyfavor The change wasn't incorporated into Flutter until [v1.16.3](https://github.com/flutter/flutter/commit/144eb45aa447bd7abde4f05cb24c6c13ced0971f). Please see the tag in that roll commit. Just to be sure something weird didn't happen I just verified locally that that change does not exist in v1.12.13.
> 
> > The three cases in this issue still produce background OpenGL calls.
> 
> Are you testing with v1.12.13 or v1.16.3?
> 
> edit: `v1.12.13+hotfix.8` does contain the fix. Previously you mentioned that version but sometimes `v1.12.13` was mentioned. I'll see if i can reproduce the problem locally.

Yes, we are using v1.12.13.hotfix.8 in our product, and this version does contain the fix.

OK, I will test your latest fix for `Rasterizer in background`.

**Comment by MrLittleStrong** [comment]
> > The engine we use is v1.12.13 which already contains this commit. The three cases in this issue still produce background OpenGL calls.
> 
> @wqyfavor The change wasn't incorporated into Flutter until [v1.16.3](https://github.com/flutter/flutter/commit/144eb45aa447bd7abde4f05cb24c6c13ced0971f). Please see the tag in that roll commit. Just to be sure something weird didn't happen I just verified locally that that change does not exist in v1.12.13.
> 
> > The three cases in this issue still produce background OpenGL calls.
> 
> Are you testing with v1.12.13 or v1.16.3?
> 
> edit: `v1.12.13+hotfix.8` does contain the fix. Previously you mentioned that version but sometimes `v1.12.13` was mentioned. I'll see if i can reproduce the problem locally.

We are using v1.17.1  stable. However, this crash is still reported.

**Comment by wqyfavor** [comment]
> > I believe that's because the GPUDisabledSyncSwitch is false by default and it does not get chance to be changed to true when user quickly switches to background after App starts up. Actually I logged in -[FlutterEngine applicationWillResignActive]. It is not executed when App crashes so the switch is not changed to true.
> 
> This a risky change. Crashing and not-crashing looks exactly the same in the case where a user backgrounds the app before `applicationBecameActive`. It has zero effect on user experience. I wouldn't recommend any change for that.

Your latest fix about Rasterizer is OK.

By the way, I mentioned that in our temporary fix for this problem, we change `is_gpu_disabled_sync_switch_` to true by default in Shell.cc. You are right. That's not correct and dangerous.
So we revert the change to Shell.cc and only change this line to protect OpenGL on iOS.




<img width="1177" alt="截屏2020-05-25 下午7 35 27" src="https://user-images.githubusercontent.com/113776/82809661-2e3e5300-9ebf-11ea-82d6-ccffa4c15855.png">

**Comment by SpiciedCrab** [comment]
So any good news from this issue?
Since our app has three flutter pages on the homepage's tab, and the crash rate increased extremely higher than before.
And also we upgrade the lates flutter version 1.17 but it doesn't work...

**Comment by gaaclarke** [comment]
@SpiciedCrab https://github.com/flutter/engine/pull/18563 is now in review.  @wqyfavor confirmed that it fixed their crash.

**Comment by gaaclarke** [comment]
fixed in https://github.com/flutter/engine/pull/18563

**Comment by SpiciedCrab** [comment]
> fixed in [flutter/engine#18563](https://github.com/flutter/engine/pull/18563)

so is it included in the latest flutter version : 1.17.3 ?

**Comment by MrLittleStrong** [comment]
> > fixed in [flutter/engine#18563](https://github.com/flutter/engine/pull/18563)
> 
> so is it included in the latest flutter version : 1.17.3 ?

not yet.

**Comment by pawlowskim** [comment]
Our users also started to complain about something like that. We switched version of flutter from 1.12.13+hotfix.8 to 1.14.3.
Crash from TestFlight (and couple very similar):
```
Thread 0 Crashed:
0   libobjc.A.dylib               	0x000000019a117d90 objc_release + 16 (objc-object.h:174)
1   Flutter                       	0x0000000103a6e630 GrMtlGpu::submitCommandBuffer(GrMtlGpu::SyncQueue) + 3008048 (GrMtlGpu.mm:184)
2   Flutter                       	0x0000000103a6df04 GrMtlGpu::onFinishFlush(GrSurfaceProxy**, int, SkSurface::BackendSurfaceAccess, GrFlushInfo const&, GrPrepareForExternalIORequests const&) + 3006212 (GrMtlGpu.mm:216)
3   Flutter                       	0x000000010398a5f4 GrGpu::finishFlush(GrSurfaceProxy**, int, SkSurface::BackendSurfaceAccess, GrFlushInfo const&, GrPrepareForExternalIORequests const&) + 2074100 (GrGpu.cpp:0)
4   Flutter                       	0x0000000103983b94 GrDrawingManager::flush(GrSurfaceProxy**, int, SkSurface::BackendSurfaceAccess, GrFlushInfo const&, GrPrepareForExternalIORequests const&) + 2046868 (GrDrawingManager.cpp:397)
5   Flutter                       	0x0000000103983f18 GrDrawingManager::flushSurfaces(GrSurfaceProxy**, int, SkSurface::BackendSurfaceAccess, GrFlushInfo const&) + 2047768 (GrDrawingManager.cpp:526)
6   Flutter                       	0x00000001039a2728 GrRenderTargetContext::flush(SkSurface::BackendSurfaceAccess, GrFlushInfo const&) + 2172712 (GrRenderTargetContext.cpp:2175)
7   Flutter                       	0x0000000103a1b7f4 SkGpuDevice::flush() + 2668532 (SkGpuDevice.cpp:1141)
8   Flutter                       	0x00000001037a4b40 flutter::FlutterPlatformViewsController::SubmitFrame(GrContext*, std::__1::shared_ptr<flutter::IOSContext>, SkCanvas*) + 84800 (FlutterPlatformViews.mm:477)
9   Flutter                       	0x00000001037bb578 flutter::IOSSurface::SubmitFrame(GrContext*, SkCanvas*) + 177528 (ios_surface.mm:0)
10  Flutter                       	0x000000010380e628 flutter::Rasterizer::DrawToSurface(flutter::LayerTree&) + 517672 (rasterizer.cc:353)
11  Flutter                       	0x000000010380ef50 std::__1::__function::__func<flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_0, std::__1::allocator<flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_0>, void (std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >)>::operator()(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&) + 520016 (functional:1651)
12  Flutter                       	0x000000010380e8d0 flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >) + 518352 (rasterizer.cc:127)
13  Flutter                       	0x00000001038142e4 std::__1::__function::__func<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_29, std::__1::allocator<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_29>, void ()>::operator()() + 541412 (functional:1651)
14  Flutter                       	0x00000001037ccc90 fml::MessageLoopImpl::FlushTasks(fml::FlushType) + 248976 (message_loop_impl.cc:129)
15  Flutter                       	0x00000001037ce908 fml::MessageLoopDarwin::OnTimerFire(__CFRunLoopTimer*, fml::MessageLoopDarwin*) + 256264 (message_loop_darwin.mm:76)
16  CoreFoundation                	0x000000019a364190 __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 32 (CFRunLoop.c:1766)
17  CoreFoundation                	0x000000019a363ea8 __CFRunLoopDoTimer + 936 (CFRunLoop.c:2357)
18  CoreFoundation                	0x000000019a36351c __CFRunLoopDoTimers + 280 (CFRunLoop.c:2512)
19  CoreFoundation                	0x000000019a35e274 __CFRunLoopRun + 1624 (CFRunLoop.c:0)
20  CoreFoundation                	0x000000019a35d8f4 CFRunLoopRunSpecific + 480 (CFRunLoop.c:3192)
21  GraphicsServices              	0x00000001a4774604 GSEventRunModal + 164 (GSEvent.c:2246)
22  UIKitCore                     	0x000000019e531358 UIApplicationMain + 1944 (UIApplication.m:4823)
23  Runner                        	0x00000001021f765c main + 96 (AppDelegate.swift:7)
24  libdyld.dylib                 	0x000000019a1d92dc start + 4
```

Is there any prediction when the fix will be on stable channel? It's holding us from production release :/

**Comment by Vadaski** [comment]
Is it included in the latest flutter version : 1.20.x ?

**Comment by MrLittleStrong** [comment]
> Is it included in the latest flutter version : 1.20.x ?

yes
